power_grid_model_io_native_c/power_grid_model_io_native/include/power_grid_model_io_native/common/exception.hpp

```diff
@@ -0,0 +1,81 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+
+#include "common.hpp"
+#include "enum.hpp"
+
+#include <concepts>
+#include <exception>
+#include <sstream>
+#include <string>
+
+namespace power_grid_model_io_native {
+namespace detail {
+inline auto to_string(std::floating_point auto x) {
+    std::ostringstream sstr{}; // NOLINT(misc-const-correctness) // https://github.com/llvm/llvm-project/issues/57297
+    sstr << x;
+    return sstr.str();
+}
+inline auto to_string(std::integral auto x) { return std::to_string(x); }
+} // namespace detail
+
+class PGMIOError : public std::exception {
+  public:
+    void append_msg(std::string_view msg) { msg_ += msg; }
+    char const* what() const noexcept final { return msg_.c_str(); }
+
+  private:
+    std::string msg_;
+};
+
+class InvalidArguments : public PGMIOError {
+  public:
+    struct TypeValuePair {
+        std::string name;
+        std::string value;
+    };
+
+    template <std::same_as<TypeValuePair>... Options>
+    InvalidArguments(std::string const& method, std::string const& arguments) {
+        append_msg(method + " is not implemented for " + arguments + "!\n");
+    }
```

Comment:
Why template the first constructor? I don't think it adds anything to it, only possible confusion.


```suggestion
    InvalidArguments(std::string const& method, std::string const& arguments) {
        append_msg(method + " is not implemented for " + arguments + "!\n");
    }
```

Reply:
Copy/pasted from PGM. I remember struggling a bit with making it user friendly

Reply:
I understand, but here the template of the first constructor shouldn't be needed for it to keep working the same, right?